src/plugins/data/public/search/search_service.test.ts

```diff
@@ -49,26 +54,168 @@ describe('Search service', () => {
+            index: 'sample-01-rollup',
+            node: 'VFTFJxpHSdaoiGxJFLSExQ',
+            reason: {
+              type: 'illegal_argument_exception',
+              reason:
+                'Field [kubernetes.container.memory.available.bytes] of type' +
+                ' [aggregate_metric_double] is not supported for aggregation [percentiles]',
+            },
+          },
+        ];
+        shards = { total: 4, successful: 2, skipped: 0, failed: 2, failures };
+        const adapter = new RequestAdapter();
+        inspector = getMockInspector({ adapter });
+        callback = jest.fn(() => false);
+      });
+
+      it('no notifications', () => {
+        const responder = inspector.adapter.start('request1');
+        shards = { total: 4, successful: 4, skipped: 0, failed: 0 };
+        responder.ok(getMockResponseWithShards(shards));
+        data.showWarnings(inspector.adapter, callback);
+
+        expect(notifications.toasts.addWarning).toBeCalledTimes(0);
+      });
+
+      it('notifications if no callback is provided', () => {
+        const responder = inspector.adapter.start('request1');
+        responder.ok(getMockResponseWithShards(shards));
+        data.showWarnings(inspector.adapter);
+
+        expect(notifications.toasts.addWarning).toBeCalledTimes(1);
+        expect(notifications.toasts.addWarning).toBeCalledWith({
+          title: '2 of 4 shards failed',
+          text: expect.any(Function),
+        });
+      });
+
+      it('no notifications when all warnings are filtered out', () => {
+        callback = () => true;
+        const responder = inspector.adapter.start('request1');
+        responder.ok(getMockResponseWithShards(shards));
+        data.showWarnings(inspector.adapter, callback);
+
+        expect(notifications.toasts.addWarning).toBeCalledTimes(0);
+      });
+
+      it('able to single notification when some warnings are filtered', () => {
+        callback = (warning) => warning.type === 'illegal_argument_exception';
+        shards.failures = [
+          {
+            reason: {
+              type: 'illegal_argument_exception',
+              reason: 'reason of "illegal_argument_exception"',
+            },
+          },
+          {
+            reason: {
+              type: 'other_kind_of_exception',
+              reason: 'reason of other_kind_of_exception',
+            },
+          },
+          { reason: { type: 'fatal_warning', reason: 'this is a fatal warning message' } },
+        ] as unknown as estypes.ShardFailure[];
+
+        const responder = inspector.adapter.start('request1');
+        responder.ok(getMockResponseWithShards(shards));
+        data.showWarnings(inspector.adapter, callback);
+
+        expect(notifications.toasts.addWarning).toBeCalledTimes(1);
+        expect(notifications.toasts.addWarning).toBeCalledWith({
+          title: '2 of 4 shards failed',
+          text: expect.any(Function),
+        });
+      });
+
+      it('timed_out warning', () => {
```

Comment:
add a new test for timeout + shard warning. Two notifications should show

Reply:
we want to add this test even though we're not sure right now if it's possible for a response to contain shard warning failures + timeout warnings -- especially for async search. To simplify the requirements for this PR, we are going to continue what seems to be the existing behavior, and leave further refactoring/removal of the timeout handling code path for future tasks:

Reply:
[284f272](https://github.com/elastic/kibana/pull/138342/commits/284f2724d2fafbf10c98aac3e7b930132fc3f377)